File: Algorithm/segmentTree.py
'''
세그먼트 트리
https://velog.io/@kineqwer1123/Python%EB%B0%B1%EC%A4%80-2042-%EC%84%B8%EA%B7%B8%EB%A8%BC%ED%8A%B8-%ED%8A%B8%EB%A6%AC 참고함
'''

def init(start, end, node):
    if start==end:
        tree[node]=a[start]
        return tree[node]

    mid=(start+end)//2
    tree[node]=min(init(start, mid, node*2),init(mid+1, end, node*2+1))
    return tree[node]

def summit(start, end, node, left, right):
    if left>end or right<start: return 0
    if left<=start and end<=right: return tree[node]
    mid=(start+end)//2
    return min(summit(start, mid, node*2, left, right), summit(mid+1, end, node*2+1, left, right))

def update(start, end, node, index, dif):
    if index<start or index>end: return
    tree[node]=min(tree[node], dif)
    if start==end: return
    mid(start+end)//2
    update(start, mid, node*2, index, dif)
    update(mid+1, end, node*2+1, index, dif)

n=10
a=[]
tree=[0]*(4*n)
for i in range(n):
    a.append(i)

init(0, n-1, 1)
print(tree)

File: Algorithm/segmentTree.cpp
// 세그먼트 트리

#include <bits/stdc++.h>
using namespace std;

int tree[12*4];
int arr[12]={1, 9, 3, 8, 4, 5, 5, 9, 10, 3, 4, 5};

int init(int s, int e, int node){
    if(s==e) return tree[node]=arr[s]; 
    return tree[node]=init(s, (s+e)/2, node*2)+init((s+e)/2+1, e, node*2+1);
}

int sum(int s, int e, int node, int l, int r){
    if(l>e || r<s) return 0;
    if(l<=s && e<=r) return tree[node];
    return sum(s, (s+e)/2, node*2, l, r)+sum((s+e)/2+1, e, node*2+1, l, r);
}

void update(int s, int e, int node, int idx, int val){
    if(idx<s || idx>e) return;
    tree[node]+=val;
    if(s==e) return;
    update(s, (s+e)/2, node*2, idx, val);
    update((s+e)/2+1, e, node*2+1, idx, val);
}

int main(void){
    init(0, 12, 1);
    cout<<sum(0, 12, 1, 0, 12)<<"\n";
    update(0, 12, 1, 5, -5);
    cout<<sum(0, 12, 1, 0, 12)<<"\n";
}